connect.go

```diff
@@ -17,80 +17,96 @@
 package edgedb
 
 import (
-	"context"
 	"fmt"
+	"sync"
 
-	"github.com/edgedb/edgedb-go/protocol/buff"
-	"github.com/edgedb/edgedb-go/protocol/message"
+	"github.com/edgedb/edgedb-go/internal/buff"
+	"github.com/edgedb/edgedb-go/internal/message"
 	"github.com/xdg/scram"
 )
 
-func (c *baseConn) connect(ctx context.Context, cfg *connConfig) error {
-	buf := buff.New(nil)
-	buf.BeginMessage(message.ClientHandshake)
-	buf.PushUint16(0) // major version
-	buf.PushUint16(8) // minor version
-	buf.PushUint16(2) // number of parameters
-	buf.PushString("database")
-	buf.PushString(cfg.database)
-	buf.PushString("user")
-	buf.PushString(cfg.user)
-	buf.PushUint16(0) // no extensions
-	buf.EndMessage()
-
-	if err := c.writeAndRead(ctx, buf.Unwrap()); err != nil {
+func (c *baseConn) connect(r *buff.Reader, cfg *connConfig) error {
+	c.writer.BeginMessage(message.ClientHandshake)
+	c.writer.PushUint16(0) // major version
+	c.writer.PushUint16(8) // minor version
+	c.writer.PushUint16(2) // number of parameters
+	c.writer.PushString("database")
+	c.writer.PushString(cfg.database)
+	c.writer.PushString("user")
+	c.writer.PushString(cfg.user)
+	c.writer.PushUint16(0) // no extensions
+	c.writer.EndMessage()
+
+	if err := c.writer.Send(c.conn); err != nil {
 		return err
 	}
 
-	for buf.Next() {
-		switch buf.MsgType {
+	var (
+		err  error
+		once sync.Once
+	)
+
+	doneReadingSignal := make(chan struct{}, 1)
+	done := func() { doneReadingSignal <- struct{}{} }
+
+	for r.Next(doneReadingSignal) {
+		switch r.MsgType {
 		case message.ServerHandshake:
 			// The client _MUST_ close the connection
 			// if the protocol version can't be supported.
 			// https://edgedb.com/docs/internals/protocol/overview
-			major := buf.PopUint16()
-			minor := buf.PopUint16()
+			major := r.PopUint16()
+			minor := r.PopUint16()
 
 			if major != 0 || minor != 8 {
-				if err := c.conn.Close(); err != nil {
-					return err
-				}
-
-				return fmt.Errorf(
-					"unsupported protocol version: %v.%v",
-					major,
-					minor,
+				e := fmt.Errorf(
```

Comment:
While here I'd make the min (or max?) supported protocol version a constant.

Reply:
done :heavy_check_mark: